nlp_architect/utils/metrics.py

```diff
@@ -94,3 +122,409 @@ def tagging(preds, labels):
+        >>> get_entities(seq)
+        [('PER', 0, 1), ('LOC', 3, 3)]
+    """
+    # for nested list
+    if any(isinstance(s, list) for s in seq):
+        seq = [item for sublist in seq for item in sublist + ["O"]]
+
+    prev_tag = "O"
+    prev_type = ""
+    begin_offset = 0
+    chunks = []
+    for i, chunk in enumerate(seq + ["O"]):
+        if suffix:
+            tag = chunk[-1]
+            type_ = chunk.split("-")[0]
+        else:
+            tag = chunk[0]
+            type_ = chunk.split("-")[-1]
+
+        if end_of_chunk(prev_tag, tag, prev_type, type_):
+            chunks.append((prev_type, begin_offset, i - 1))
+        if start_of_chunk(prev_tag, tag, prev_type, type_):
+            begin_offset = i
+        prev_tag = tag
+        prev_type = type_
+
+    return chunks
+
+
+def end_of_chunk(prev_tag, tag, prev_type, type_):
+    """Checks if a chunk ended between the previous and current word.
+
+    Args:
+        prev_tag: previous chunk tag.
+        tag: current chunk tag.
+        prev_type: previous type.
+        type_: current type.
+
+    Returns:
+        chunk_end: boolean.
+    """
+    chunk_end = False
+
+    end_tag = ("L", "E")
+    start_tag = ("U", "S")
+
+    if prev_tag in end_tag:
+        chunk_end = True
+    if prev_tag in start_tag:
+        chunk_end = True
+
+    if prev_tag == "B" and tag == "B":
+        chunk_end = True
+    if prev_tag == "B" and tag in start_tag:
+        chunk_end = True
+    if prev_tag == "B" and tag == "O":
+        chunk_end = True
+    if prev_tag == "I" and tag == "B":
+        chunk_end = True
+    if prev_tag == "I" and tag in start_tag:
+        chunk_end = True
+    if prev_tag == "I" and tag == "O":
+        chunk_end = True
+
+    if prev_tag != "O" and prev_tag != "." and prev_type != type_:
+        chunk_end = True
+
+    return chunk_end
+
+
+def start_of_chunk(prev_tag, tag, prev_type, type_):
+    """Checks if a chunk started between the previous and current word.
+
+    Args:
+        prev_tag: previous chunk tag.
+        tag: current chunk tag.
+        prev_type: previous type.
+        type_: current type.
+
+    Returns:
+        chunk_start: boolean.
+    """
+    chunk_start = False
+
+    end_tag = ("L", "E")
+    start_tag = ("U", "S")
+
+    if tag == "B":
+        chunk_start = True
+    if tag in start_tag:
+        chunk_start = True
+
+    if prev_tag in end_tag and tag in end_tag:
+        chunk_start = True
+    if prev_tag in end_tag and tag == "I":
+        chunk_start = True
+    if prev_tag in start_tag and tag in end_tag:
+        chunk_start = True
+    if prev_tag in start_tag and tag == "I":
+        chunk_start = True
+    if prev_tag == "O" and tag in end_tag:
+        chunk_start = True
+    if prev_tag == "O" and tag == "I":
+        chunk_start = True
+
+    if tag != "O" and tag != "." and prev_type != type_:
+        chunk_start = True
+
+    return chunk_start
+
+
+def f1_score(y_true, y_pred, suffix=False):
```

Comment:
Please rename entity metrics to reflect that these are NER metrics (f1 and accuracy exist for other tasks and it might be confusing)

Reply:
done